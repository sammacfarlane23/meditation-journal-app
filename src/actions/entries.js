import database from '../firebase/firebase';

export const createEntry = (entry) => ({
  type: 'CREATE_ENTRY',
  entry,
});

export const startCreateEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { entryText = '', createdAt = 0, title = '' } = entryData;
    const entry = {
      entryText,
      createdAt,
      title,
    };

    return database
      .ref(`users/${uid}/entries`)
      .push(entry)
      .then((ref) => {
        dispatch(
          createEntry({
            id: ref.key,
            ...entry,
          })
        );
      });
  };
};

export const removeEntry = ({ id } = {}) => ({
  type: 'REMOVE_ENTRY',
  id,
});

export const startRemoveEntry = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .remove()
      .then(() => {
        dispatch(removeEntry({ id }));
      });
  };
};

export const editEntry = (id, updates) => ({
  type: 'EDIT_ENTRY',
  id,
  updates,
});

export const startEditEntry = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editEntry(id, updates));
      });
  };
};

export const setEntries = (entries) => {
  return {
    type: 'SET_ENTRIES',
    entries,
  };
};

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries`)
      .once('value')
      .then((snapshot) => {
        const entries = [];

        snapshot.forEach((childSnapshot) => {
          entries.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setEntries(entries));
      });
  };
};

export const selectEntryForEdit = (entry) => {
  return {
    type: 'SELECT_ENTRY_FOR_EDIT',
    entry,
  };
};
