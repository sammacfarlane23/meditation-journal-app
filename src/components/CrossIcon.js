import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';

export default (props) => (
  <div>
    <button className='footer-button' onClick={props.showSearch}>
      <FontAwesomeIcon icon={faTimes} className='white-button' size='2x' />
    </button>
  </div>
);
