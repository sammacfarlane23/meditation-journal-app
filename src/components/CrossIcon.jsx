import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';

export default (props) => (
  <div>
    <button className='button--big' onClick={props.showSearch}>
      <FontAwesomeIcon icon={faTimes} size='2x' />
    </button>
  </div>
);
