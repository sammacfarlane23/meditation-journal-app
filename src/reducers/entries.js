const entriesReducerDefaultState = [];

export default (state = entriesReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_ENTRY':
      return [...state, action.entry];
    case 'REMOVE_ENTRY':
      return state.filter(({ id }) => action.id !== id);
    case 'EDIT_ENTRY':
      return state.map((entry) => {
        if (entry.id === action.id) {
          return {
            ...entry,
            ...action.updates,
          };
        } else {
          return entry;
        }
      });
    case 'SET_ENTRIES':
      return action.entries;
    case 'SELECT_ENTRY_FOR_EDIT':
      return action.entry;
    default:
      return state;
  }
};
