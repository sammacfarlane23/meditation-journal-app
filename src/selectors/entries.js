export default (entries, { text }) => {
  return entries
    .filter((entry) => {
      const titleMatch = entry.title.toLowerCase().includes(text.toLowerCase());

      const textMatch = entry.entryText
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch || titleMatch;
    })
    .sort((a, b) => {
      return a.createdAt < b.createdAt ? 1 : -1;
    });
};
