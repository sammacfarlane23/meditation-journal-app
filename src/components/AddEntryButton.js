import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus } from '@fortawesome/free-solid-svg-icons';
import EntryModal from './EntryModal';
import { startCreateEntry } from '../actions/entries';

export default () => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const dispatch = useDispatch();

  const showModal = () => {
    setIsModalOpen(true);
  };

  const onSubmit = (entry) => {
    dispatch(startCreateEntry(entry));
    setIsModalOpen(false);
  };

  const closeModal = () => {
    setIsModalOpen(false);
  };

  return (
    <div className='button-position'>
      {!isModalOpen && (
        <button className='button--big circular-container' onClick={showModal}>
          <FontAwesomeIcon icon={faPlus} size='2x' />
        </button>
      )}
      <EntryModal
        showModal={isModalOpen}
        onSubmit={onSubmit}
        closeModal={closeModal}
      />
    </div>
  );
};
