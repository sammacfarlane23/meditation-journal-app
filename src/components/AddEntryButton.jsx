import React from 'react';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus } from '@fortawesome/free-solid-svg-icons';
import EntryModal from './EntryModal';
import { startCreateEntry } from '../actions/entries';

export class AddEntryButton extends React.Component {
  state = { show: false };

  showModal = () => {
    this.setState(() => ({ show: true }));
  };

  onSubmit = (entry) => {
    this.props.startCreateEntry(entry);
    this.setState({ show: false });
  };

  closeModal = () => {
    this.setState({ show: false });
  };
  render() {
    return (
      <div className='button-position'>
        {!this.state.show && (
          <button
            className='button--big circular-container'
            onClick={this.showModal}
          >
            <FontAwesomeIcon icon={faPlus} size='2x' />
          </button>
        )}
        <EntryModal
          showModal={this.state.show}
          onSubmit={this.onSubmit}
          closeModal={this.closeModal}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startCreateEntry: (entry) => dispatch(startCreateEntry(entry)),
});

export default connect(undefined, mapDispatchToProps)(AddEntryButton);
