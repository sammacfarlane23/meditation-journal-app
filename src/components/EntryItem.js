import React from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import EntryModal from './EntryModal';
import { startEditEntry, startRemoveEntry } from '../actions/entries';

export class EntryItem extends React.Component {
  state = {
    show: false,
    canEdit: true,
  };

  showModal = () => {
    this.setState(() => ({ show: true }));
  };

  onRemove = (entry) => {
    this.props.startRemoveEntry({ id: this.props.entry.id });
    this.setState({ show: false });
  };

  onSubmit = (entry) => {
    this.props.startEditEntry(this.props.entry.id, entry);
    this.setState({ show: false });
  };

  closeModal = () => {
    this.setState({ show: false });
  };
  render() {
    return (
      <div className='entry-item'>
        <button
          className='entry-item entry-item--button'
          onClick={this.showModal}
        >
          <div className='entry-list-title'>
            <h1 className='entry-list-title'>
              {moment(this.props.createdAt).format('DD MMM YYYY HH:mm')}
            </h1>
          </div>
          {this.props.title && (
            <h1 className='entry-item-title'>{this.props.title}</h1>
          )}
          <div className='entry-list-text'>{this.props.entryText}</div>
        </button>
        <EntryModal
          showModal={this.state.show}
          selectedEntry={this.props.entry}
          onSubmit={this.onSubmit}
          closeModal={this.closeModal}
          canEdit={this.state.canEdit}
          onRemove={this.onRemove}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    entry: state.entries.find((entry) => entry.id === props.id),
  };
};

const mapDispatchToProps = (dispatch) => ({
  startEditEntry: (id, entry) => dispatch(startEditEntry(id, entry)),
  startRemoveEntry: (entry) => dispatch(startRemoveEntry(entry)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EntryItem);
