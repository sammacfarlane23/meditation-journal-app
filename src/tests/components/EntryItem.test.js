import React from 'react';
import { shallow } from 'enzyme';
import { EntryItem } from '../../components/EntryItem';
import entries from '../fixtures/entries';

test('should render EntryItem correctly without title', () => {
  const wrapper = shallow(<EntryItem {...entries[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render EntryItem correctly with title', () => {
  const wrapper = shallow(<EntryItem {...entries[1]} />);
  expect(wrapper).toMatchSnapshot();
});
