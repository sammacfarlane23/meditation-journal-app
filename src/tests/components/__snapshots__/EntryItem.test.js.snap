// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EntryItem correctly with title 1`] = `
<div>
  <button
    className="entry-item"
    onClick={[Function]}
  >
    <div>
      <h1
        className="entry-item__date"
      >
        01 Jan 1970 00:59
      </h1>
    </div>
    <h1
      className="entry-item__title"
    >
      title
    </h1>
    <div
      className="entry-item__text"
    >
      This is another entry
    </div>
  </button>
  <Connect(EntryModal)
    canEdit={true}
    closeModal={[Function]}
    onRemove={[Function]}
    onSubmit={[Function]}
    showModal={false}
  />
</div>
`;

exports[`should render EntryItem correctly without title 1`] = `
<div>
  <button
    className="entry-item"
    onClick={[Function]}
  >
    <div>
      <h1
        className="entry-item__date"
      >
        01 Jan 1970 01:00
      </h1>
    </div>
    <div
      className="entry-item__text"
    >
      This is an entry
    </div>
  </button>
  <Connect(EntryModal)
    canEdit={true}
    closeModal={[Function]}
    onRemove={[Function]}
    onSubmit={[Function]}
    showModal={false}
  />
</div>
`;
